teps/0015-pause-pipeline.md

```diff
@@ -0,0 +1,105 @@
+---
+title: pause-pipeline
+authors:
+  - "@jbarrick-mesosphere"
+creation-date: 2020-09-10
+last-updated: 2020-09-10
+status: draft
+---
+
+# TEP-0015: Support Paused PipelineRuns and TaskRuns
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Platform Implementor](#platform-implementor)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a method for pausing PipelineRuns and TaskRuns that are then
+ignored by the Tekton reconciler. These paused resources can be used by platform
+implementers to control PipelineRun and TaskRun resources, for example, in cases
+where the cluster is under heavy load and not ready to process a PipelineRun.
+
+PipelineRuns and TaskRuns can be created paused so that they are not started at
+all or, in the case of a PipelineRun, the pause setting can be used to prevent
+further TaskRuns from being created. As it is not possible to "pause" a
+Kubernetes pod, the pause setting would not affect TaskRuns after they are
+started.
+
+## Motivation
+
+The primary motivation for this is to support operators who need to control the
+start up of PipelineRuns and TaskRuns in clusters that are under heavy load.
+By allowing PipelineRuns to be created without actually starting them, one
+could implement a custom controller to implement any PipelineRun scheduling
```

Comment:
One can already implement custom scheduling today at the Kubernetes level, which would let you prioritize Pod startup, which is roughly equivalent to TaskRun startup.

I think I'd want to see more motivation about why we should implement this at the Tekton level as opposed to at the Kubernetes level, where there's already been a ton of work done to support basically anything you might want.

Reply:
We do want to use the advanced scheduling options in Kubernetes, in addition to some other pipeline-aware features.

My primary use-case here is to allow us to create a parallelism limit for PipelineRuns. We will also utilize pod priority to order the tasks in the pipelines that do run, which will help somewhat (also this is not something we can do until Kubernetes 1.19, as currently it is not possible to disable pre-emption). But ordering of pods that run and limiting the actual pipelines that are submitted to the scheduler are separate use-cases.

The issue that we are trying to solve is that when you create a new PipelineRun, if there are many other PipelineRuns with complex pipelines you can have many PipelineRuns competing over the same resources - causing the scheduler to be saturated and for individual tasks to take a very long time to start.

For example, if we were to use ResourceQuotas to limit the number of pods in a namespace, we only limit the number of _tasks_ that can run simultaneously, not the number of _pipelines_.

So we want to be able to limit the number of PipelineRuns that are started at any given time. We do not want to implement the actual parallelism limit at the Tekton level because we want to have custom limits for the PipelineRuns (e.g., based on the namespace or for a particular repository's pipelines), so this feature allows us to submit inert PipelineRuns and then have Tekton act on them once we're ready.

Reply:
Perhaps this design is more complex than I intended and kind of obfuscates my intentions. The only feature that I need here is to be able to create inert PipelineRuns. The rest about pausing TaskRuns and supporting pausing after the PipelineRun has already been started is only because those features seemed natural to include at the same time for consistency/least surprise.

Reply:
Yeah, let's start with the minimal use case you want to add then when we get to an agreement about it we can discuss how to extend it. Going the other way leads to unused features that were "easy to add" that add support burden.

My main question is: Why not have whatever system is creating-and-unpausing PipelineRuns just hold off on creating them until they're safe to run? TaskRuns/Pods that don't have schedulable resources will already wait until resources become available.

Reply:
Okay cool, I'll slim the scope of this document and update it.

> My main question is: Why not have whatever system is creating-and-unpausing PipelineRuns just hold off on creating them until they're safe to run? TaskRuns/Pods that don't have schedulable resources will already wait until resources become available.

I tried to address this in the alternatives section, but basically this would require us to have a non-Kubernetes-native data store (say, Redis) that would be a lot harder for our users to have visibility into or to create a CRD that is an abstraction over the PipelineRun, which we are hesitant to do since it would add quite a bit of complexity.

Reply:
Is there a reason that it's not preferrable to create the PipelineRun and just having TaskRun Pods queue until resources are available?

If you've tried that already and it didn't work, that would be very useful concrete information to include in the proposal, and might help us focus our design efforts on the specific problems you've encountered.

Reply:
If you have many pipelines running and they each are complex having many taskruns and/or dependencies, your scheduling queue will be saturated with taskruns from many pipelines causing no pipeline to proceed in a timely manner. It's very easy to cause pipeline timeouts or just very long run times in this scenario. What this allows us to do is help ensure that new pipelines aren't causing extra queue saturation - and we also will have multiple vectors for this such as limiting the number of pipelines created by a particular project at any given time to help limit queue pressure.

We will eventually be able to take advantage of pod priorities to help order our queue (when Kubernetes v1.19 comes out), but we also want to limit the pipelines actually being submitted to the scheduling queue. For example, if I have a bunch of little jobs and a big job, I might want to limit the big job to one at a time so it's not taking up lots of queue space from the little jobs.

Reply:
This definitely sounds like the job for a custom scheduler, based on your description. Pods emitted by Tekton are annotated with PipelineRun information, so your scheduler could try to prioritize those belonging to already-started PipelineRuns and not assign Pods to nodes until it determines that it can make sufficient progress on it.

Depending on how complex you want to get, you could even annotate PipelineRuns with priorities and eviction options, node affinity with PVCs, the sky's the limit.

This could even serve as a useful proving ground for custom scheduling features we might want to incorporate into Tekton itself, based on your own usage and gained experience building it outside of Tekton first. By not building it into Tekton you'd also have more flexibility to explore without being saddled by our compatibility guarantees and review requirements and release schedules.

Reply:
No. I have looked into custom scheduling before and it's quite complex. You would essentially have to re-implement the entire Kubernetes scheduler or fork it to be able to reproduce all of the advanced scheduling features that Kubernetes already supports.

See https://thenewstack.io/implementing-advanced-scheduling-techniques-with-kubernetes/ to understand exactly what you're suggesting would entail.

As an example of this, check out kube-batch. It is _incredibly_ difficult to debug because they don't produce errors exactly like the regular kubernetes scheduler. And it's an official Kubernetes sig project.

All we need here is the ability to create inert pipelines and then the rest of our use-case can be handled externally by our own integration. Besides that, this feature can be useful for other things, for example, scheduled jobs.

Essentially, all this feature is is a short circuit for the Tekton reconciler.

Reply:
It might help me to understand what kind of system you're planning to build to update "paused" PipelineRuns to "running", since this is likely where you'll end up writing all the interesting scheduling code if you decide not to put it in a Kubernetes scheduler. What criteria would it use to determine when a PipelineRun should be unpaused?

Reply:
It will just be a simple PipelineRun controller that watches PipelineRuns and starts new ones when old ones finish.

Something like:

* Notified of PipelineRun changes.
* If the PipelineRun is finished, check which PipelineRuns are pending. If it is a pending PipelineRun, we'll use that.
* If there is a pending PipelineRun, check to see if there are more PipelineRuns running than the parallelism limit, if not, start the PipelineRun.

The "criteria" would be user configured parallelism limits.

Reply:
That makes sense, and is roughly what I'd imagined you were going for, so that's good. 👍 

Here's an alternative:

1. Add a ConfigMap option for `concurrent-pipelineruns-per-namespace` or something similar (not sure if your requirements are per-namespace or per-cluster, this is just an example).
1. Have the PipelineRun controller consult this config when it sees a new PipelineRun, and if there are already >=N PipelineRuns going, it sets its `.status.conditions[type=Succeeded]=Unknown` ("ongoing"), with `Reason=ConcurrencyCapped` (naming tbd), and doesn't create any TaskRuns.
1. When a PipelineRun completes, the PipelineRun controller identifies a `ConcurrencyCapped` PipelineRun ... somehow ... and sets it to `Running`, and proceeds to create TaskRuns as today, etc.

This has some benefits as I see it:


- It clearly identifies the feature as "concurrency caps", and not "general-purpose pausing", which I think helps with scoping.
- Concurrency caps are (I think) a primarily concern of the Operator, not of the User or any external system. This model makes concurrency caps an operator-level option with all the others in a ConfigMap. If an Operator doesn't want to have a cap, they configure that in the ConfigMap, and no external system can be installed that can change that.
- Paused-ness is a `status`, and not expressed in a `spec` -- paused-ness decisions are a black-box to the User and Operator, and Tekton retains the ability to change how that works without breaking user expectations (modulo [Hyrum's Law](https://www.hyrumslaw.com/), of course)
- This doesn't introduce a new integration point for an external system (the "unpauser") -- it brings pausing/unpausing in-house. This avoids the potential situation where possibly multiple pauser/unpausers are operating at once on the same pool of PipelineRuns -- that way lies madness. 🐉 

And some disadvantages:

- Whether concurrency caps are enforced in Tekton or outside, there are going to be races when a PipelineRun is unblocked. This is basically unavoidable, but under this model it's Tekton's responsibility.
- In this model, pausedness is _even more_ integrated into Tekton; the pausing/unpausing logic is in Tekton's reconciler.
- Related to the above, any changes to capping behavior has to go through Tekton's review process, release schedules, etc. -- Tekton's contributors and community are responsible for any bugs.

Whether this is preferrable, I think it's worth talking through its relative merits and including it as an option in the proposal.

Reply:
Like I said previously, we need to be able to do our own custom limits, not related strictly to PipelineRun and don't think it's a good idea to include this logic in Tekton itself as this is unnecessary complexity in Tekton.

> Have the PipelineRun controller consult this config when it sees a new PipelineRun, and if there are already >=N PipelineRuns going, it sets its .status.conditions[type=Succeeded]=Unknown ("ongoing"), with Reason=ConcurrencyCapped (naming tbd), and doesn't create any TaskRuns.

If this is supported already I'm certainly happy to use it. If it's not, I'm not too particular on the actual API we decide on for the feature I need.

> This doesn't introduce a new integration point for an external system (the "unpauser") -- it brings pausing/unpausing in-house. This avoids the potential situation where possibly multiple pauser/unpausers are operating at once on the same pool of PipelineRuns -- that way lies madness. dragon

This is a _disadvantage_ not a benefit.

Reply:
> Like I said previously, we need to be able to do our own custom limits, not related strictly to PipelineRun

What kinds of limits are you considering? It sounded like from https://github.com/tektoncd/community/pull/203#discussion_r487232538 it was user-defined concurrency limits (I realize now this presumably means end-users, not Operators, that changes requirements a bit)

> If this is supported already I'm certainly happy to use it.

This _is_ actually possible today, I think. You could have a mutating admission webhook force new PipelineRuns into the `Succeeded=Unknown` state (so Tekton doesn't try to start them immediately) and a reconciling controller with permission to update PipelineRun statuses to push them into a running state, at which point Tekton's normal reconciliation would take over and start any TaskRuns.

This _should_ be prototypeable today, though I don't think anybody's done anything like it yet. If you're interested in exploring it I'd be happy to help, and if you find any bugs while investigating I'd be happy to fix them.

> This is a disadvantage not a benefit.

Eye of the beholder I suppose 😆 . The thought of multiple systems separately judging the same pool of PipelineRuns seems like a real mess to me.

Reply:
> This is actually possible today, I think. You could have a mutating admission webhook force new PipelineRuns into the Succeeded=Unknown state (so Tekton doesn't try to start them immediately) and a reconciling controller with permission to update PipelineRun statuses to push them into a running state, at which point Tekton's normal reconciliation would take over and start any TaskRuns.

Yeah, I'll try this out and report back. Thanks!

Reply:
> What kinds of limits are you considering? It sounded like from #203 (comment) it was user-defined concurrency limits (I realize now this presumably means end-users, not Operators, that changes requirements a bit)

Depends on use cases, the granularity of applying the limits may not be aligned with namespaces. For example, some may want to set concurrency limits per service account or per "repository."

Reply:
It doesn't look like it supports this status yet as it immediately runs:

```
status:
  conditions:
  - lastTransitionTime: "2020-09-15T18:15:29Z"
    message: 'Not started yet.'
    reason: ConcurrencyCapped
    status: "Unknown"
    type: Succeeded
```

I went ahead and narrowed the scope of the document to just PipelineRuns. I didn't change the API interface yet, but if everyone prefers it as a condition I am happy to change it to that.

Reply:
Actually it looks like #3223 uses the `PipelineRunPaused` setting to tell Tekton to pause, which sets the Status condition to Paused. I'll change the TEP to that.